packages/x-data-grid-pro/src/hooks/features/serverSideLazyLoader/useGridDataSourceLazyLoader.ts

```diff
@@ -0,0 +1,483 @@
+import * as React from 'react';
+import { throttle } from '@mui/x-internals/throttle';
+import {
+  useGridApiEventHandler,
+  useGridSelector,
+  gridSortModelSelector,
+  gridFilterModelSelector,
+  GridEventListener,
+  GRID_ROOT_GROUP_ID,
+  GridGroupNode,
+  GridSkeletonRowNode,
+  gridPaginationModelSelector,
+  gridDimensionsSelector,
+  gridFilteredSortedRowIdsSelector,
+  useFirstRender,
+} from '@mui/x-data-grid';
+import {
+  getVisibleRows,
+  GridGetRowsParams,
+  gridRenderContextSelector,
+  GridStrategyProcessor,
+  useGridRegisterStrategyProcessor,
+} from '@mui/x-data-grid/internals';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+import { findSkeletonRowsSection } from '../lazyLoader/utils';
+import { GRID_SKELETON_ROW_ROOT_ID } from '../lazyLoader/useGridLazyLoaderPreProcessors';
+import { DataSourceRowsUpdateStrategy, runIf } from '../dataSource/utils';
+
+enum LoadingTrigger {
+  VIEWPORT,
+  SCROLL_END,
+}
+
+const INTERVAL_CACHE_INITIAL_STATE = {
+  firstRowToRender: 0,
+  lastRowToRender: 0,
+};
+
+const getSkeletonRowId = (index: number) => `${GRID_SKELETON_ROW_ROOT_ID}-${index}`;
+
+/**
+ * @requires useGridRows (state)
+ * @requires useGridPagination (state)
+ * @requires useGridDimensions (method) - can be after
+ * @requires useGridScroll (method
+ */
+export const useGridDataSourceLazyLoader = (
+  privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    | 'pagination'
+    | 'paginationMode'
+    | 'unstable_dataSource'
+    | 'lazyLoading'
+    | 'lazyLoadingRequestThrottleMs'
+    | 'scrollEndThreshold'
+  >,
+): void => {
+  const setStrategyAvailability = React.useCallback(() => {
+    privateApiRef.current.setStrategyAvailability(
+      'dataSource',
+      DataSourceRowsUpdateStrategy.LazyLoading,
+      props.unstable_dataSource && props.lazyLoading ? () => true : () => false,
+    );
+  }, [privateApiRef, props.lazyLoading, props.unstable_dataSource]);
+
+  const [lazyLoadingRowsUpdateStrategyActive, setLazyLoadingRowsUpdateStrategyActive] =
+    React.useState(false);
+  const sortModel = useGridSelector(privateApiRef, gridSortModelSelector);
+  const filterModel = useGridSelector(privateApiRef, gridFilterModelSelector);
+  const paginationModel = useGridSelector(privateApiRef, gridPaginationModelSelector);
+  const filteredSortedRowIds = useGridSelector(privateApiRef, gridFilteredSortedRowIdsSelector);
+  const dimensions = useGridSelector(privateApiRef, gridDimensionsSelector);
+  const renderContext = useGridSelector(privateApiRef, gridRenderContextSelector);
```

Comment:
I think that it should not be a problem for this hook, since it does not need super fast response. Context is used to generate the new request params but it has throttle on it anyway

Reply:
Sounds good, we can still do changes as the API is unstable. 👍